### DISCUSSION TITLE
The PL Translation vs the fheroes2 UI

### DISCUSSION THREAD

**Discussion by Arthusppp** [ihhub/fheroes2#4486] [General]
Hello all. I am raising the following: I am struggling with the PL translation, please see the below screenshots.

What I want to discuss here: how do we want to tackle this? Or this is me and my resolution?
Do we have similar problems in other languages?

**Example#1:**
The fheroes2 presents the unit stats in a more clear way. However, when the Polish text is longer, the text exceeds the dialog...

fheroes2:
![fheroes2_3eflee5qHn](https://user-images.githubusercontent.com/85434237/138571557-eb13d1df-31dc-44cd-9060-8c0479920bea.png)

OG:
![Heroes2w_0uhyiT3GTz](https://user-images.githubusercontent.com/85434237/138571552-85dce7b4-bd38-428b-ba2c-2179f3b0d3ca.png)

---

**Example#2:** 
The Polish version of the OG cheats here! Instead of the "Obrażenia", the OG uses "Obr"; and instead of "Punkty życia" it uses "PŻ"... What approach should we take? Should we do the same? 
And one more "cheat", please see the "Mage tower": the OG puts its name into two lines... What here? I believe we should do the same, right?
Lastly, the OG scales somehow units: please compare Golem with Giant... In the OG's well, the height of the Golem and Giant is the same. The consequence is that in the fheroes2, Giant does not fit its cell...

fheroes2:
![fheroes2_XBMMhpCqRv](https://user-images.githubusercontent.com/85434237/138571583-7085f477-9b9d-4f01-ab4d-3c6578099030.png)

OG:
![Heroes2w_MK5Mn4SbGG](https://user-images.githubusercontent.com/85434237/138571594-24197910-3949-4308-ab67-85ff848c104d.png)

---
**Example#3:** 
Please see the last box in the top row. fheroes2 has "Truność mapy" in two lines, but horizontally centered. However, the OG found strange workaround: first two boxes uses two lines, but last uses one line...
Secondly, in the race name row, we should change the display a bit (cut off too long names).
I believe both are small enhancements in the fheroes2, right?

fheroes2:
![fheroes2_QCKl2fQmpx](https://user-images.githubusercontent.com/85434237/138571745-6c15f376-f041-4430-bddc-b82eddc1c3ce.png)

OG:
![Heroes2w_hizSrs1CRS](https://user-images.githubusercontent.com/85434237/138571732-3ff3d053-1636-4237-a283-2a4dc4dadf95.png)

---
Example#4
How to tackle the secondary skill level?

fheroes2:
![fheroes2_1dLt11m9Vl](https://user-images.githubusercontent.com/85434237/138572458-f6b3422e-222a-4f5e-b0f6-77cc8c2ff5b8.png)

OG:
![Heroes2w_nRPlr3ZIXV](https://user-images.githubusercontent.com/85434237/138572462-353778c0-0729-49cd-b738-4f9718f442c2.png)

**Comment by Branikolog** [comment]
Hello, @Arthusppp !

> Lastly, the OG scales somehow units: please compare Golem with Giant... In the OG's well, the height of the Golem and Giant is the same. The consequence is that in the fheroes2, Giant does not fit its cell...

You can notice, that in the OG Golem is placed a bit higher. But still, Giant's helmet is still higher a bit.
In fheroes2 all sprites are placed with the same space from the bottom of the box and in relation to their bottom.

> **Example#1:** The fheroes2 presents the unit stats in a more clear way. However, when the Polish text is longer, the text exceeds the dialog...
> 
> fheroes2: ![fheroes2_3eflee5qHn](https://user-images.githubusercontent.com/85434237/138571557-eb13d1df-31dc-44cd-9060-8c0479920bea.png)

Well, Talking about shots, we can easily take away first word... 
But I really don't have an ideas how could morale be shortened....

**Reply by zenseii** [reply]
Hello @Arthusppp and @Branikolog, I've been looking into these exact things lately in order to solve these text wrapping issues.  I can confirm that these problems are present in other languages too. 

Issue 1 - long stat names and descriptions:
The way I solved this in the Norwegian translation was sadly to use different shorter words that were not as good as what I would have chosen. This is especially valid for both Speed and Morale since if you look at a unit in combat it will also have an additional "(x)" after the string to show an integer representation of the units level of speed. This means that for the text to not go outside of the window the word needs to be 9 characters or shorter iirc, which is really short.

Looking at this now I realise a solution might be to allow to have multiple lines for these stats. So that a longer word would break to the next line.

One thing I would have to check in that case is how many lines we have at disposal since I think in some windows there are options or stuff beneath the stats. I believe that in army and castle view you have the options to dismiss, upgrade and cancel, but there should still be room for at least two extra lines.

Issue 2 - Mage tower in well textbox:
Making mage tower description into two lines should be easy, but how does this look when you've actually bought the mage tower and the number of recruitable mages are shown?

Issue 3 - Alignment of game parameters and long class names for the scenarioinfo and gameinfo windows:

I currently have a PR going in order to solve the issues with the multiple lines for game parameter headers PR #4482 :
Here's what it looks like right now (ignore the missing letters with accents, of course, and the multiple-lined hero class names):

![image](https://user-images.githubusercontent.com/12501091/139197264-61a35925-775c-41a5-bd57-e0f834a89431.png)


Hero classes overlapping:

I've been looking a lot into this one and I am not sure how to solve this. Adding an additional line is easy to do and helps a bit. However, this would make long class names get divided at random places like ex in the Norwegian translation Sorceress gets split like this:
![image](https://user-images.githubusercontent.com/12501091/139192859-fc8625b8-c218-4d20-a8e2-fca8d9526f99.png)
when the correct way would be to use the dash:

> Troll-
> kjerring

The problem is that you cannot change the name of the classes to add a dash and a space because this affects class names all over the game so like in headers in the Hero window or when recruiting a hero. This is apart from Necromancer which currently has its own unique "Necroman" string to match the English OG. 

One possibility could be to define separate strings for every class to be used in the context of these windows. Something like "class|Knight", "class|Sorceress". etc.  

Another problem is that having multiple lines means that the swap player "double-arrowed button" in scenario info window when selecting a map gets too close to the second line of the class name. You can move it down but things might start to feel a bit cluttered. Not sure if you could maybe increase the height of the window.


Issue 4 - Secondary skill level:
I'm not sure how to solve that one. You could make it multiple lines but that would maybe cover up too much of the icon. Also, why isn't the same translation as the OG used here (I don't know Polish so you'll have to excuse me)?

**Reply by Branikolog** [reply]
Hello, @zenseii.

> Looking at this now I realise a solution might be to allow to have multiple lines for these stats. So that a longer word would break to the next line.
> 
> One thing I would have to check in that case is how many lines we have at disposal since I think in some windows there are options or stuff beneath the stats. I believe that in army and castle view you have the options to dismiss, upgrade and cancel, but there should still be room for at least two extra lines.

I'm afraid, that you've forgotten that we also have spells icons here with durability during a battle. Which takes lots of place. And close\dismiss\upgrade buttons in the bottom. Therefore, we don't have enough space to add 1 more text line, especially for ranged troops, as they have one characteristic more, than others.

The only solution I can imagine here is to change the words, so they fit the line, or come back to OG alignment of these characteristics, which is not so well distinctive...

> Hero classes overlapping:
>
> I've been looking a lot into this one and I am not sure how to solve this. Adding an additional line is easy to do and helps a bit.

I'm not sure, if it would look well, but we can shift all icons down a little and allow engine draw one class-name below a portrait, the second - above, the third below etc.
Or shift 2nd, 4th and 6th class-names a little down, so they don't overlap with 1st, 3rd and 5th...
(Only for the case with 6 players)


> Issue 4 - Secondary skill level:
> I'm not sure how to solve that one. You could make it multiple lines but that would maybe cover up too much of the icon. Also, why isn't the same translation as the OG used here (I don't know Polish so you'll have to excuse me)?

We can use OG names or use abbreviation with dot in the end...

**Reply by zenseii** [reply]
> I'm afraid, that you've forgotten that we also have spells icons here with durability during a battle. Which takes lots of place. And close\dismiss\upgrade buttons in the bottom. Therefore, we don't have enough space to add 1 more text line, especially for ranged troops, as they have one characteristic more, than others.

That's a good point and you're right I had completely forgotten about it.... I take it increasing the height of the creature info window isn't a good idea either?
> The only solution I can imagine here is to change the words, so they fit the line, or come back to OG alignment of these characteristics, which is not so well distinctive...

So you're saying we could move the placements of the statistics like maybe align them to the left or have the line start from the left. Then there's always to consider sticking to the OG looks of things.

> I'm not sure, if it would look well, but we can shift all icons down a little and allow engine draw one class-name below a portrait, the second - above, the third below etc. Or shift 2nd, 4th and 6th class-names a little down, so they don't overlap with 1st, 3rd and 5th... (Only for the case with 6 players)

Yes the problem could be according to how many players are in the game, although I still think it is better to have the text within the limits of the borders of the portraits.

This is how two lines with 6 players would look:

![image](https://user-images.githubusercontent.com/12501091/139222298-58559c94-12a3-42a8-a627-7cf81b4ca187.png)

Not sure how having some names above the portraits and some below would look. Maybe a bit messy? And there isn't much room above the portraits either.

> We can use OG names or use abbreviation with dot in the end...

Yes, a dot is an okay solution. I had to use it for the Mage Guild in the build window.

**Reply by Branikolog** [reply]
> I take it increasing the height of the creature info window isn't a good idea either?

It is the best solution for the issue. But the problem is that we have to draw a completely new window, which is not a trivial task. (Original window is not generated by an engine, but has fixed size.)


> This is how two lines with 6 players would look:

Warlock's name with a space in a second line looks not good, from my point. :)
But not sure, it is worse, than placing names wider than a portrait, actually....

**Reply by zenseii** [reply]
> It is the best solution for the issue. But the problem is that we have to draw a completely new window, which is not a trivial task. 

Yes, that is not an easy task and should probably be the last resort then.

> Warlock's name with a space in a second line looks not good, from my point. :) But not sure, it is worse, than placing names wider than a portrait, actually....

Haha yes I think I need to have the Polish Heroes 2 version installed to actually get the Polish special letters so that's why they are all just replaced by blank spaces. I might be wrong, but this is what I get when I make the .mo.

So in your opinion it would look better to have the names go beyond the limit of the portraits? I mean the intent of the OG way of doing it is to avoid this, which is why Necromancer is simply just truncated.

**Reply by Branikolog** [reply]
> So in your opinion it would look better to have the names go beyond the limit of the portraits? I mean the intent of the OG way of doing it is to avoid this, which is why Necromancer is simply just truncated.

It's just one possible solution. I'm not sure it will look great. Also considering the fact, some of names should be shifted down....
